### DISCUSSION TITLE
Questions&Answers

### DISCUSSION THREAD

**Discussion by koculu** [koculu/ZoneTree#5] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/7453467/185395622-bed3177b-8489-4193-9905-9b25ac8ba6c3.png)

## You could ask or answer questions here.

**Comment by ibrahimakgn** [comment]
Hi,
Is there a way to store class objects at value ? I couldnt use ISerializer interface for serialize my objects. Is there any example for it?

**Reply by koculu** [reply]
Hi @thornzz,
Have you checked the second sample in the readme file?
You can pass your custom serializer through **SetValueSerializer** method.

![image](https://user-images.githubusercontent.com/7453467/199574603-cfc54c2a-5c54-4640-8637-e0ac0478aff6.png)

**Comment by ibrahimakgn** [comment]
Thanks. I have created custom serializer for my object and its works.

**Comment by aley1** [comment]
Hi @koculu 

Hope you can assist me with these questions:
1. Does ZoneTree allow setting the max number of key/value items that kept in memory? for example keep only 50,000 keys in memory and rest on disk segment.
2. Could you please provide some code example how to turn off WAL
3. How do we handle multi-threaded access to a ZT database? I presume only one thread should be writing to the db, but can other threads access the data at the same time?

Thanks.

**Reply by koculu** [reply]
Hi @aley1 
1. Please use [MutableSegmentMaxItemCount ](https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.Options.ZoneTreeOptions-2.html#Tenray_ZoneTree_Options_ZoneTreeOptions_2_MutableSegmentMaxItemCount)to limit the maximum size of mutable region via [ZoneTreeOptions](https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.Options.ZoneTreeOptions-2.html). 
For read-only region, you could use the maintainer [ThresholdForMergeOperationStart ](https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.IMaintainer.html#Tenray_ZoneTree_IMaintainer_ThresholdForMergeOperationStart)property.
The sum of these two properties controls the count of the records in memory.

2. Use [ConfigureWriteAheadLogOptions ](https://tenray.io/docs/ZoneTree/api/Tenray.ZoneTree.ZoneTreeFactory-2.html#Tenray_ZoneTree_ZoneTreeFactory_2_ConfigureWriteAheadLogOptions_Action_Tenray_ZoneTree_Options_WriteAheadLogOptions__)from factory.
```C#
factory.ConfigureWriteAheadLogOptions(x => { x.WriteAheadLogMode = WriteAheadLogMode.None; })
```
3. ZoneTree is thread-safe. You don't have to worry about synchronization at all. You could write and read from multiple threads safely.

You may find example use cases for all options in the tests and benchmark projects. Please look there for more details to not to turn this conversation into a personal advisor talk:)

**Comment by KSemenenko** [comment]
Hi, is there any way to transfer wal to another instance of zonetree?

I'm a bit stuck at the stage of synchronizing two distributed zonetree.

I initially thought to send the changes to all instances, but if I lose the message I need to somehow find in wal the position from which to add the data

**Reply by koculu** [reply]
Hi @KSemenenko 
You can create your custom Write Ahead Log and provide that to be used by ZoneTree.
On the other hand, the distribution of key-value databases is a hard problem.
You may need a consensus algorithm.
References: 
https://en.wikipedia.org/wiki/Consensus_(computer_science)
https://raft.github.io/
https://martinfowler.com/articles/patterns-of-distributed-systems/wal.html

**Comment by ljcollins25** [comment]
How do you generate the API docs for ZoneTree?

**Reply by koculu** [reply]
Build the project using `ReleaseWithDoc` configuration.

**Comment by koculu** [comment, marked as answer]
Questions & Answers page is locked.
Please create new discussion pages when you want to ask questions.
